### DISCUSSION TITLE
posts/wpf-name-vs-xname/

### DISCUSSION THREAD

**Discussion by giscus** [BYJRK/byjrk.github.io#21] [Announcements]
# posts/wpf-name-vs-xname/

在 WPF 开发中，可以给控件添加 Name 或 x:Name 属性。那么这二者究竟是什么区别呢？本文就来简单探讨一下。

https://blog.coldwind.top/posts/wpf-name-vs-xname/

**Comment by Trarizon** [comment]
我反而喜欢特地用`x:Name`，和c++的`std::`一样看起来莫名有种爽感（

**Comment by momj-00** [comment]
建议使用x:Name，不是所有类型都可以使用Name，但是任何类型都可以使用x:Name

**Comment by ZerosZhang** [comment]
`Name`属性定义在`FrameworkElement`中，这个类是`WPF`控件的基类，所以所有的`WPF`控件都具有`Name`这个属性。当一个元素具有`Name`属性时，使用`Name`或`x:Name`效果是一样的。

```C#
<Button x:Name="Button_Hello" />
<Button Name="Button_Hello" />

```

  此时`Name`和`x:Name`是可以互换的，只是不能同时出现在一个元素中。

  对于那些没有`Name`属性的元素，在`XAML`声明时也希望创建引用变量以便于在`C#`代码中进行查找，则只能使用`x:Name`。

  因此`x:Name`的功能涵盖了`Name`属性的功能，所以建议全部使用`x:Name`以增强代码的统一性和可读性。

**Reply by ZerosZhang** [reply]
想到了一些例子：
比如 `ListView` 中的 `GridView` 的 `GridViewColumn` ，并非继承自 `FrameworkElement` ，所以没有 `Name` 这个属性，但是可以通过使用 `x:Name` 来访问。同样的，`DataGrid` 的 `DataGridColumn` 也并非继承自 `FrameworkElement`，也只能使用 `x:Name` 进行访问。

**Reply by jlennonxzy** [reply]
谢谢解惑.
